handshake/src/types/tx.rs

```diff
@@ -0,0 +1,462 @@
+    fn write_sighash_preimage<W: Write>(
+        &self,
+        writer: &mut W,
+        args: &Self::SighashArgs,
+    ) -> Result<(), Self::TxError> {
+        if args.sighash_flag == Sighash::None || args.sighash_flag == Sighash::NoneACP {
+            return Err(TxError::NoneUnsupported);
+        }
+
+        if (args.sighash_flag == Sighash::Single || args.sighash_flag == Sighash::SingleACP)
+            && args.index >= self.outputs().len()
+        {
+            return Err(TxError::SighashSingleBug);
+        }
+
+        let input = &self.vin[args.index];
+
+        Self::write_u32_le(writer, self.version)?;
+        self.hash_prevouts(args.sighash_flag)?.write_to(writer)?;
+        self.hash_sequence(args.sighash_flag)?.write_to(writer)?;
+        input.outpoint.write_to(writer)?;
+        args.prevout_script.write_to(writer)?;
+        Self::write_u64_le(writer, args.prevout_value)?;
+        Self::write_u32_le(writer, input.sequence)?;
+        self.hash_outputs(args.index, args.sighash_flag)?
+            .write_to(writer)?;
+        Self::write_u32_le(writer, self.locktime)?;
+        Self::write_u32_le(writer, args.sighash_flag as u32)?;
+
+        Ok(())
+    }
+}
+
+/// An Error type for transaction objects
+#[derive(Debug, Error)]
+pub enum TxError {
+    /// Serialization-related errors
+    #[error(transparent)]
+    SerError(#[from] SerError),
+
+    /// IOError bubbled up from a `Write` passed to a `ByteFormat::serialize` implementation.
+    #[error(transparent)]
+    IOError(#[from] IOError),
+
+    /// Sighash NONE is unsupported
+    #[error("SIGHASH_NONE is unsupported")]
+    NoneUnsupported,
+
+    /// Satoshi's sighash single bug. Throws an error here.
+    #[error("SIGHASH_SINGLE bug is unsupported")]
+    SighashSingleBug,
+
+    /// Caller provided an unknown sighash type to `Sighash::from_u8`
+    #[error("Unknown Sighash: {}", .0)]
+    UnknownSighash(u8),
+
+    /// Got an unknown flag where we expected a witness flag. May indicate a non-witness
+    /// transaction.
+    #[error("Witness flag not as expected. Got {:?}. Expected {:?}.", .0, [0u8, 1u8])]
+    BadWitnessFlag([u8; 2]),
+
+    /// Wrong sighash args passed in to wrapped tx
+    #[error("Sighash args must match the wrapped tx type")]
+    WrongSighashArgs,
+    // /// No outputs in vout
+    // #[error("Vout may not be empty")]
+    // EmptyVout
+}
+
+/// Type alias for result with TxError
+pub type TxResult<T> = Result<T, TxError>;
+
+/// Signature hash args
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct SighashArgs {
+    /// The index of the input we'd like to sign
+    pub index: usize,
+    /// The sighash mode to use.
+    pub sighash_flag: Sighash,
+    /// The script used in the prevout, which must be signed. In complex cases involving
+    /// `OP_CODESEPARATOR` this must be the subset of the script containing the `OP_CHECKSIG`
+    /// currently being executed.
+    /// TODO: create a script type
+    pub prevout_script: Vec<u8>,
```

Comment:
Need a `Script` type to use, the scripts in handshake are more like bytecode since its segwit only, they never exist bare in scriptPubKeys. If the data in the address is 20 bytes, its assumed to be a public key and templated into a script. It differs from bitcoin though because it uses blake160